File: reporting/export_report.mjs
import fs from 'fs';
import path from 'path';
import crypto from 'crypto';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

function sha256File(p) {
  const b = fs.readFileSync(p);
  return crypto.createHash('sha256').update(b).digest('hex');
}

function sampleRecords() {
  return [
    { id: 1, titolo: 'Bug login', risolto: false },
    { id: 2, titolo: 'Aggiornare policy', risolto: true },
    { id: 3, titolo: 'Errore pagamento', risolto: false }
  ];
}

function parseArgs() {
  const argv = process.argv.slice(2);
  const out = { outDir: path.join(__dirname, 'output'), sessionId: null, dataFile: null };
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === '--out-dir' || a === '-o') out.outDir = argv[++i];
    else if (a === '--session-id' || a === '-s') out.sessionId = argv[++i];
    else if (a === '--data-file' || a === '-i') out.dataFile = argv[++i];
  }
  return out;
}

async function renderReport(records, outDir, sessionId) {
  fs.mkdirSync(outDir, { recursive: true });
  const totale = records.length;
  const risolti = records.filter(r => r.risolto).length;
  const nonRisolti = totale - risolti;
  const timestamp = new Date().toISOString().replace(/[:.-]/g, '').slice(0, 15);
  const filenameBase = `report_${timestamp}_${sessionId}`;

  // CSV
  const csvPath = path.join(outDir, `${filenameBase}.csv`);
  const csvLines = [];
  csvLines.push(['ID', 'Titolo', 'Risolto'].join(','));
  records.sort((a, b) => (a.risolto === b.risolto) ? a.id - b.id : (a.risolto ? 1 : -1)).forEach(r => {
    csvLines.push([r.id, `"${String(r.titolo).replace(/"/g, '""')}"`, r.risolto ? 'Risolto' : 'Non risolto'].join(','));
  });
  fs.writeFileSync(csvPath, csvLines.join('\r\n'), { encoding: 'utf8' });

  // Chart (optional)
  let chartPath = null;
  try {
    const { ChartJSNodeCanvas } = await import('chartjs-node-canvas');
    const width = 400; const height = 400;
    const chartCallback = (ChartJS) => {
      // noop for now
    };
    const charter = new ChartJSNodeCanvas({ width, height, chartCallback });
    const configuration = {
      type: 'pie',
      data: {
        labels: ['Risolti', 'Non risolti'],
        datasets: [{ data: [risolti, nonRisolti], backgroundColor: ['#4CAF50', '#F44336'] }]
      },
      options: { plugins: { legend: { position: 'bottom' } } }
    };
    const buffer = await charter.renderToBuffer(configuration);
    chartPath = path.join(outDir, `${filenameBase}_chart.png`);
    fs.writeFileSync(chartPath, buffer);
  } catch (e) {
    console.warn('Chart generation skipped:', e.message || e);
  }

  // PDF
  let pdfPath = path.join(outDir, `${filenameBase}.pdf`);
  try {
    const PDFDocument = (await import('pdfkit')).default;
    const doc = new PDFDocument({ margin: 40 });
    const ws = fs.createWriteStream(pdfPath);
    doc.pipe(ws);

    doc.fontSize(16).font('Helvetica-Bold').text('Report KPI Segnalazioni', { align: 'center' });
    doc.moveDown();

    doc.fontSize(12).font('Helvetica').text(`Totale segnalazioni: ${totale}`);
    doc.text(`Risolti: ${risolti}`);
    doc.text(`Non risolti: ${nonRisolti}`);
    doc.moveDown();

    if (chartPath && fs.existsSync(chartPath)) {
      try { doc.image(chartPath, { fit: [300, 300], align: 'center' }); doc.moveDown(); } catch (e) { console.warn('Unable to add chart to PDF:', e.message || e); }
    }

    // Table header
    const startX = doc.x;
    const cellHeight = 20;
    doc.font('Helvetica-Bold');
    doc.rect(doc.x, doc.y, 50, cellHeight).stroke(); doc.text('ID', doc.x + 5, doc.y + 5, { width: 50 - 10 });
    doc.rect(doc.x + 50, doc.y, 300, cellHeight).stroke(); doc.text('Titolo', doc.x + 55, doc.y + 5, { width: 300 - 10 });
    doc.rect(doc.x + 350, doc.y, 100, cellHeight).stroke(); doc.text('Stato', doc.x + 355, doc.y + 5, { width: 100 - 10 });
    doc.moveDown();

    doc.font('Helvetica');
    records.sort((a, b) => (a.risolto === b.risolto) ? a.id - b.id : (a.risolto ? 1 : -1)).forEach(r => {
      const y = doc.y;
      doc.rect(doc.x, y, 50, cellHeight).stroke(); doc.text(String(r.id), doc.x + 5, y + 5, { width: 50 - 10 });
      doc.rect(doc.x + 50, y, 300, cellHeight).stroke(); doc.text(String(r.titolo), doc.x + 55, y + 5, { width: 300 - 10 });
      doc.rect(doc.x + 350, y, 100, cellHeight).stroke();
      if (r.risolto) doc.fillColor('#008000'); else doc.fillColor('#C80000');
      doc.text(r.risolto ? 'Risolto' : 'Non risolto', doc.x + 355, y + 5, { width: 100 - 10 });
      doc.fillColor('black');
      doc.moveDown();
    });

    doc.end();
    // wait for stream finish
    await new Promise((res, rej) => ws.on('finish', res).on('error', rej));
  } catch (e) {
    console.warn('PDF generation failed:', e.message || e);
    // ensure pdfPath points to something or null
    if (!fs.existsSync(pdfPath)) pdfPath = null;
  }

  // meta
  const meta = { run_id: sessionId, generated_at: new Date().toISOString(), csv: path.basename(csvPath), pdf: pdfPath ? path.basename(pdfPath) : null };
  try { meta.csv_sha256 = sha256File(csvPath); if (pdfPath) meta.pdf_sha256 = sha256File(pdfPath); } catch (e) { meta.hash_error = String(e); }

  const metaPath = path.join(outDir, `${filenameBase}.meta.json`);
  fs.writeFileSync(metaPath, JSON.stringify(meta, null, 2), { encoding: 'utf8' });

  return { csv: csvPath, pdf: pdfPath, meta: metaPath };
}

async function main() {
  const opts = parseArgs();
  const outDir = path.isAbsolute(opts.outDir) ? opts.outDir : path.join(process.cwd(), opts.outDir);
  const sessionId = opts.sessionId || crypto.randomBytes(4).toString('hex');
  let records = [];
  if (opts.dataFile) {
    try { records = JSON.parse(fs.readFileSync(opts.dataFile, 'utf8')); } catch (e) { console.warn('Unable to read data file, using sample:', e.message); records = sampleRecords(); }
  } else records = sampleRecords();

  const artifacts = await renderReport(records, outDir, sessionId);
  console.log('Report CSV salvato in:', artifacts.csv);
  console.log('Report PDF salvato in:', artifacts.pdf);
  console.log('Meta salvato in:', artifacts.meta);
}

if (process.argv[1].endsWith('export_report.mjs')) main().catch(e => { console.error(e); process.exit(1); });

File: reporting/export_report.py
#!/usr/bin/env python3
timestamp = datetime.datetime.now().strftime("%Y%m%d_%H%M%S")
session_id = "ABC123"  # In produzione: generare dinamicamente
import argparse
import csv
import datetime
import hashlib
import json
import uuid
from pathlib import Path
from fpdf import FPDF
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt


def sha256_file(p: Path):
    h = hashlib.sha256()
    with p.open('rb') as f:
        for chunk in iter(lambda: f.read(8192), b''):
            h.update(chunk)
    return h.hexdigest()


def load_sample_records():
    return [
        {"id": 1, "titolo": "Bug login", "risolto": False},
        {"id": 2, "titolo": "Aggiornare policy", "risolto": True},
        {"id": 3, "titolo": "Errore pagamento", "risolto": False}
    ]


def render_report(records, out_dir: Path, session_id: str):
    out_dir.mkdir(parents=True, exist_ok=True)
    totale = len(records)
    risolti = sum(1 for r in records if r["risolto"])
    non_risolti = totale - risolti

    timestamp = datetime.datetime.now().strftime("%Y%m%d_%H%M%S")
    filename_base = f"report_{timestamp}_{session_id}"

    # CSV
    csv_path = out_dir / f"{filename_base}.csv"
    with open(csv_path, "w", newline="", encoding="utf-8") as csvfile:
        writer = csv.writer(csvfile)
        writer.writerow(["ID", "Titolo", "Risolto"])
        for r in sorted(records, key=lambda x: x["risolto"]):
            writer.writerow([r["id"], r["titolo"], "Risolto" if r["risolto"] else "Non risolto"])

    # Grafico (pie)
    plt.figure(figsize=(4, 4))
    colors = ["#4CAF50", "#F44336"]
    plt.pie([risolti, non_risolti], labels=["Risolti", "Non risolti"], autopct="%1.1f%%", colors=colors)
    plt.title("Stato segnalazioni")
    chart_path = out_dir / f"{filename_base}_chart.png"
    plt.savefig(chart_path, bbox_inches='tight')
    plt.close()

    # PDF
    class PDF(FPDF):
        def header(self):
            self.set_font("Arial", "B", 14)
            self.cell(0, 10, "Report KPI Segnalazioni", 0, 1, "C")

        def footer(self):
            self.set_y(-15)
            self.set_font("Arial", "I", 8)
            self.cell(0, 10, f"Pagina {self.page_no()}", 0, 0, "C")

    pdf = PDF()
    pdf.set_auto_page_break(auto=True, margin=15)
    pdf.add_page()

    # KPI in PDF
    pdf.set_font("Arial", "", 12)
    pdf.cell(0, 10, f"Totale segnalazioni: {totale}", ln=True)
    pdf.cell(0, 10, f"Risolti: {risolti}", ln=True)
    pdf.cell(0, 10, f"Non risolti: {non_risolti}", ln=True)
    pdf.ln(5)

    # Grafico nel PDF
    try:
        pdf.image(str(chart_path), x=60, w=80)
    except Exception as e:
        print('Warning: unable to add chart to PDF:', e)
    pdf.ln(10)

    # Tabella segnalazioni
    pdf.set_font("Arial", "B", 12)
    pdf.cell(20, 10, "ID", 1)
    pdf.cell(100, 10, "Titolo", 1)
    pdf.cell(40, 10, "Stato", 1)
    pdf.ln()

    pdf.set_font("Arial", "", 12)
    for r in sorted(records, key=lambda x: x["risolto"]):
        stato = "Risolto" if r["risolto"] else "Non risolto"
        pdf.cell(20, 10, str(r["id"]), 1)
        pdf.cell(100, 10, r["titolo"], 1)
        if r["risolto"]:
            pdf.set_text_color(0, 128, 0)
        else:
            pdf.set_text_color(200, 0, 0)
        pdf.cell(40, 10, stato, 1)
        pdf.set_text_color(0, 0, 0)
        pdf.ln()

    # Salvataggio PDF
    pdf_path = out_dir / f"{filename_base}.pdf"
    pdf.output(str(pdf_path))

    # meta
    meta = {
        "run_id": session_id,
        "generated_at": datetime.datetime.now().isoformat(),
        "csv": str(csv_path.name),
        "pdf": str(pdf_path.name),
    }
    try:
        meta["csv_sha256"] = sha256_file(csv_path)
        meta["pdf_sha256"] = sha256_file(pdf_path)
    except Exception as e:
        meta["hash_error"] = str(e)

    meta_path = out_dir / f"{filename_base}.meta.json"
    with open(meta_path, 'w', encoding='utf-8') as mf:
        json.dump(meta, mf, indent=2, ensure_ascii=False)

    return {"csv": csv_path, "pdf": pdf_path, "meta": meta_path}


def main():
    parser = argparse.ArgumentParser(description='Export report CSV+PDF with KPI and chart')
    parser.add_argument('--out-dir', '-o', default=str(Path(__file__).resolve().parent / 'output'))
    parser.add_argument('--session-id', '-s', default=None)
    parser.add_argument('--data-file', '-i', default=None, help='optional JSON file with records')
    args = parser.parse_args()

    out_dir = Path(args.out_dir)
    session_id = args.session_id or uuid.uuid4().hex[:8]

    # load data
    if args.data_file:
        import json
        with open(args.data_file, 'r', encoding='utf-8') as df:
            records = json.load(df)
    else:
        records = load_sample_records()

    artifacts = render_report(records, out_dir, session_id)
    print(f"Report CSV salvato in: {artifacts['csv']}")
    print(f"Report PDF salvato in: {artifacts['pdf']}")
    print(f"Meta salvato in: {artifacts['meta']}")


if __name__ == '__main__':
    main()
